harness/determined/common/api/_session.py

```diff
@@ -1,27 +1,93 @@
-from typing import Any, Dict, Optional
+import abc
+import json as _json
+from typing import Any, Dict, Optional, Tuple, Union
 
 import requests
 import urllib3
 
+import determined as det
+from determined.common import requests as det_requests
 from determined.common import util
-from determined.common.api import authentication, certs, request
+from determined.common.api import authentication, certs, errors, request
 
 
-class Session:
-    def __init__(
-        self,
-        master: Optional[str],
-        user: Optional[str],
-        auth: Optional[authentication.Authentication],
-        cert: Optional[certs.Cert],
-        max_retries: Optional[urllib3.util.retry.Retry] = None,
-    ) -> None:
-        self._master = master or util.get_default_master_address()
-        self._user = user
-        self._auth = auth
-        self._cert = cert
-        self._max_retries = max_retries
+def _do_request(
```

Comment:
the fact that this method that makes all the underlying HTTP requests is basically entirely decoupled from the "session" seems suspicious to me... like what's the point of the `Session`? it's basically just a data object that adds a token to the header, and it seems like the token store is actually where the auth/cache logic is handled. functionally it makes no difference if callers of `Session` reuse the same one or create a new one if the token is the same.

Reply:
i was expecting the Session to be something that is designed to be reused for multiple HTTP requests. technically it is i suppose, but really only the cert/token get passed in. and it seems like we don't even reuse the same underlying HTTP session each request, we make a new HTTP session every time. 

i see that you didn't actually change anything about this logic, so i don't know if this is out of scope for this PR. but why aren't we reusing HTTP connections in these sessions? pretty sure this would non-trivially improve performance, and now we already have this session object that all HTTP requests hang off of. `Session.__init__` can initialize and store the HTTP connection/session (`requests.Session` or `urllib3.ConnectionPool` i think), then the `_do_request` calls can use it?

Reply:
This sounds like a great refactor you could do after this PR lands.

Seriously though, I eliminated probably more than a hundred calls to the old do_request() function, and now everything goes through a Session.  This refactor you are proposing sounds like a compelling idea but I think it should be step 2 where this PR is step 1.  Your proposal would be impossible without this refactor first.